Add troubleshooting section for Elastic Agent upgrade failing on Windows (#1750)
elastic/docs-content: troubleshoot/ingest/fleet/common-problems.md

The error is caused by Windows [desktop heap exhaustion](https://learn.microsoft.com/en-us/troubleshoot/windows-server/performance/desktop-heap-limitation-out-of-memory). When {{agent}} runs as a [Windows service application](https://learn.microsoft.com/en-us/dotnet/framework/windows-services/introduction-to-windows-service-applications), it uses the service desktop, and shares the desktop heap with other running services. If a service process is using windowing resources, but is failing to release them, this may exhaust the desktop heap and affect {{agent}}.

:::{note}
Interactively-run instances of `elastic-agent.exe` are not subject to this limitation. Only instances running as a service are potentially affected.
:::

To resolve the issue, you can try the following:

- **Update {{agent}} immediately after a system reboot**

>     A system reboot will destroy and recreate the desktop heap, resolving any prior exhaustion. Since many memory leaks are gradual, updating {{agent}} immediately after a system reboot may allow {{agent}} to upgrade before the memory leaking application exhausts the desktop heap.

Reviewer: ```suggestion
    A system reboot destroys and recreates the desktop heap, resolving any prior exhaustion. 
    Because many memory leaks are gradual, updating {{agent}} immediately after a system reboot may allow {{agent}} to upgrade before the memory leaking application exhausts the desktop heap.
```

    :::{tip}

>     A [cold startup](https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/distinguishing-fast-startup-from-wake-from-hibernation) will reset kernel memory, but a fast startup or a wake from hibernation will not. A regular reboot (for example, `shutdown /r /t 0`) will result in a cold startup, and will reset the desktop heap.

Reviewer: ```suggestion
    A [cold startup](https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/distinguishing-fast-startup-from-wake-from-hibernation) resets kernel memory, but a fast startup or a wake from hibernation does not. 
    A regular reboot (for example, `shutdown /r /t 0`) results in a cold startup, and resets the desktop heap.
```

    :::

- **Update third-party service applications**

    As standard Windows tools such as Task Manager and Process Explorer do not attribute desktop heap usage by application, you have to consider updating all third-party processes that are running as a service. To list these applications, use the following PowerShell command:

    ```powershell
    PS C:\> Get-Process | Where {$_.SI -eq 0} | Where {$_.MainModule.FileVersionInfo.ProductName -and (-not (($_.MainModule.FileVersionInfo.CompanyName -eq "Microsoft Corporation") -and ($_.MainModule.FileVersionInfo.ProductName -like "*Windows*"))) } | ForEach-Object { $_.MainModule.FileVersionInfo.ProductName + ' - ' + $_.Path }
    ```

    You can then install any updates from the listed applications' manufacturers.

- **Terminate or uninstall third-party service applications**

>     You can try terminating or uninstalling non-critical third-party service applications before updating {{agent}}. Terminating a process will release its desktop heap resources.

Reviewer: ```suggestion
    You can try terminating or uninstalling non-critical third-party service applications before updating {{agent}}. 
    Terminating a process releases its desktop heap resources.
```
